accepted/2023/NuGetAudit-supress-advisory.md

```diff
@@ -0,0 +1,230 @@
+
+- Author: [zivkan](https://github.com/zivkan)
+- GitHub Issue: https://github.com/NuGet/Home/issues/11926
+- todo: Another issue for NuGetAudit suppressions
+
+## Summary
+
+<!-- One-paragraph description of the proposal. -->
+
+Provide a syntax that allows developers to give a list of URLs that NuGet will no longer warn about in restore (NuGetAudit) or `dotnet list package --vulnerable`.
+
+## Motivation
+
+<!-- Why are we doing this? What pain points does this solve? What is the expected outcome? -->
+Software with known vulnerabilities are often only vulnerable only in certain scenarios.
+But not all applications using a library with a known vulnerability use it in a way that's at risk.
+There are also times when upgrading is not easy.
+For example, if a package introduced a breaking change, so an application is using an old version, but the security vulnerability is only fixed in the newest version of the package, it could require non-trivial effort to upgrade the package.
+
+While it's possible to use `NoWarn` to ignore all known vulnerabilities of a specific severity, this will also prevent any future vulnerability of the same severity from being reported.
+
+An example is there have been multiple deserialization libraries that had known vulnerabilities reported where the attack is a deeply nested object hierarchy leading to a stack overflow situation.
+The recommended fix in these situations are usually to configure the deserialization library to limit the maximum depth it will allow.
+On .NET, this changes from an unrecoverable `StackOverflowException`, to a recoverable (via `try-catch`) exception, with a more meaningful error.
+However, for some apps, this doesn't matter, either because the app still fails, or because the stream being read is trusted not to have malicious content.
+In these scenarios, the development team might choose to ignore one advisory, but still wish to be informed of any new advisories for this package that may come out in the future.
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or pseudocode to show how this proposal would look. -->
+
+Add an MSBuild `NuGetAuditSuppress` item for the URL that NuGet reports. For example, to suppress https://github.com/advisories/GHSA-5crp-9r3c-p9vr, add `<NuGetAuditSuppress Include="https://github.com/advisories/GHSA-5crp-9r3c-p9vr" />`.
+
+Here is a complete project example:
+
+```xml
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFramework>net8.0</TargetFramework>
+    <OutputType>exe</OutputType>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Newtonsoft.Json" Version="9.0.1" />
+    <NuGetAuditSuppress Include="https://github.com/advisories/GHSA-5crp-9r3c-p9vr" />
```

Comment:
In general, when developers suppress code analysis warnings there is an option to provide some justification for future reference. Would it make sense to provide an option for users to add justification for suppressing these warnings as well?

Reply:
With items, users will be able to add their own justification attribute if they wish. We're not looking to provide any tooling around a "Justification" attribute, but users can still add it themselves.